Source/JavaScriptCore/wasm/WasmCallProfile.h

```diff
@@ -57,22 +61,106 @@ class CallProfile {
+                profile.m_boxedCallee = boxedCallee;
+                profile.incrementCount();
+                return;
+            }
+        }
+        makeMegamorphic();
     }
 
     EncodedJSValue boxedCallee() const { return m_boxedCallee; }
 
     static constexpr ptrdiff_t offsetOfCount() { return OBJECT_OFFSETOF(CallProfile, m_count); }
     static constexpr ptrdiff_t offsetOfBoxedCallee() { return OBJECT_OFFSETOF(CallProfile, m_boxedCallee); }
 
-    static constexpr EncodedJSValue initCallee = 0b00;
-    static constexpr EncodedJSValue polymorphicCallee = 0b01;
-    static constexpr EncodedJSValue megamorphicCallee = 0b10;
+    static constexpr EncodedJSValue initCallee = 0b0000;
+    static constexpr EncodedJSValue polymorphicCallee = 0b0100;
+    static constexpr EncodedJSValue megamorphicCallee = 0b1000;
+    static constexpr EncodedJSValue calleeMask = 0b1100;
+#if USE(JSVALUE64)
+    static_assert(!(JSValue::NativeCalleeTag & calleeMask));
+#endif
+
+    static constexpr size_t maxPolymorphicCallees = 3;
+
+    class alignas(16) PolymorphicCallee final : public TrailingArray<PolymorphicCallee, CallProfile> {
+        WTF_DEPRECATED_MAKE_FAST_ALLOCATED(PolymorphicCallee);
+        WTF_MAKE_NONMOVABLE(PolymorphicCallee);
+        using TrailingArrayType = TrailingArray<PolymorphicCallee, CallProfile>;
+        friend TrailingArrayType;
+    public:
+
+        static std::unique_ptr<PolymorphicCallee> create(unsigned size, CallProfile* profile)
+        {
+            return std::unique_ptr<PolymorphicCallee>(new (fastMalloc(allocationSize(size))) PolymorphicCallee(size, profile));
+        }
+
+        static constexpr ptrdiff_t offsetOfProfile() { return OBJECT_OFFSETOF(PolymorphicCallee, m_profile); }
+
+    private:
+        PolymorphicCallee(unsigned size, CallProfile* profile)
+            : TrailingArrayType(size)
+            , m_profile(profile)
+        {
+        }
+
+        CallProfile* m_profile { nullptr };
+    };
+
+    static bool isMegamorphic(EncodedJSValue boxedCallee)
+    {
+        return boxedCallee & megamorphicCallee;
+    }
+
+    static Callee* monomorphic(EncodedJSValue boxedCallee)
+    {
+        if (boxedCallee & megamorphicCallee)
+            return nullptr;
+        if (boxedCallee & polymorphicCallee)
+            return nullptr;
+        uintptr_t bits = static_cast<uintptr_t>(boxedCallee & ~calleeMask);
+        if (!bits)
+            return nullptr;
+        return std::bit_cast<Callee*>(CalleeBits(bits).asNativeCallee());
+    }
+
+    static PolymorphicCallee* polymorphic(EncodedJSValue boxedCallee)
+    {
+        if (boxedCallee & megamorphicCallee)
+            return nullptr;
+        if (boxedCallee & polymorphicCallee)
+            return std::bit_cast<PolymorphicCallee*>(static_cast<uintptr_t>(boxedCallee & ~calleeMask));
+        return nullptr;
+    }
 
 private:
+    void makeMegamorphic()
+    {
+        m_boxedCallee = (m_boxedCallee | megamorphicCallee);
```

Comment:
When you make it megamorphic, do you want to keep the polymorphic flag set? It looks like it's not used since `polymorphic()` bails if the megamorphic flag is set.